### PR TITLE
Set visibility compiler flag to reduce executable size

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

In my tests this setting reduced binary sizes by roughly 10%. Has anyone tried it and perhaps done more thorough tests with it?

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsdeppe** [comment]
It doesn't seem to build, so that's bad :man_shrugging:

**Comment by nilsvu** [comment]
Works on macOS.. perhaps older compilers have problems with it. I'll do more tests.

**Comment by nilsdeppe** [comment]
Okay, thanks. I read up a bit on the flag and it does sound like it would be nice if we can use it :)

**Comment by nilsvu** [comment]
I looked into this again to see if we can use it to reduce executable sizes, with mixed results. Here's a comparison to [this](https://github.com/sxs-collaboration/spectre/runs/1295228100) recent CI build on develop:

 Build variant | develop | this PR | Reduction in size
--- | --- | --- | ---
clang-10 Debug | bin 1.4G, lib 2.0G | bin 1.4G, lib 2.0G | none
clang-10 Release | bin 1.1G, lib 579M | bin 1.1G, lib 580M | none
gcc-9 Debug | bin 2.8G, lib 2.5G | bin 1.4G, lib 2.5G | bin 50%, lib none
gcc-9 Release | bin 1.3G, lib 554M | bin 1.1G, lib 553M | bin 15%, lib none

Looks like clang binaries don't reduce in size, but gcc binaries get down to clang's size.

These are builds with static libraries. I'm not sure yet how to handle shared libs. To hide symbols in shared libs we may have to annotate the source code, i.e. set the default visibility to "hidden" and explicitly un-hide public symbols. This can be done with `#pragma GCC visibility push(default) ... #pragma GCC visibility pop` or with `__attribute__((__visibility__("default")))`, possibly wrapped in a macro like `SPECTRE_PUBLIC`. Does anyone know more about best practices here? Here's an article I found linked in many places: https://gcc.gnu.org/wiki/Visibility. Do you think the reduction in binary size, load times and possibly runtime (I haven't benchmarked that, but articles like the one linked before suggest improved code optimization) are worth adding visibility annotations to our libraries?

**Comment by wthrowe** [comment]
It is still not really clear to me what this does or how to use it.  The article spends a lot of time touting the benefits but not very much explaining the practicalities.

The size reduction is certainly interesting, although I don't understand where it is coming from.  The article only mentions size reduction of the symbol table, and it looks like that is less than 3% of our executable size.  (This is not something I'm familiar with, so I might have screwed up this measurement.)

Does this have any effect on running the code through a debugger?

**Comment by nilsdeppe** [comment]
I agree with @wthrowe that without understanding what this we shouldn't go ahead and use it. However, given the improvements in executable size it might be worth the effort. Having said that, this should be compared with shared libs as well. The other major concerning this is regarding exceptions. The option parser uses exceptions and going from "Problem encountered on line X, column Y..." to "Execution terminated, uncaught exception" is a major regression in user friendliness. 

I'm also not generally fond of the idea of having to have a lot of annotations on symbols. This will make the source code even more complicated. Templates are bad enough, and those at least have a lot of articles all over the internet (and books) explaining what they are.  If it's only some option parsing classes (ideally the option internal ones) then I think annotations would be fine. If every class and enum that is option-creatable needs annotations, then I think we would need to get some idea of how many that is and get feedback at the spectre call.

In short: this looks super promising and I think worth understanding. However, I think we need to answer the questions @wthrowe and I have raised before we can make an informed decision on how to proceed.

**Comment by nilsdeppe** [comment]
Looking into this more there is at least the issue that C++ requires all functions (including inline functions) have the same address in every translation unit https://en.cppreference.com/w/cpp/language/inline. This means we shouldn't do `set(CMAKE_VISIBILITY_INLINES_HIDDEN ON)` because it causes us to violate standard C++. See https://stackoverflow.com/questions/48621251/why-fvisibility-inlines-hidden-is-not-the-default Interestingly, apparently Clang violates some ELF requirement, which can be done with GCC using the `-fno-semantic-interposition`.

As you can see by the link failures in: https://github.com/sxs-collaboration/spectre/pull/1945/checks?check_run_id=1302434259
 we need to annotate _every single symbol_ that is an interface as a public (default) symbol. This is _a lot_ of symbols. I expect the size reduction to be less drastic (though still drastic) if we do that, but I'm certainly not volunteering myself to annotate the hundreds to thousands of symbols we have. We will ultimately need to do something similar when we switch to C++20 and modules, but there we can do one library at a time. One library at a time might be possible with this as well, but given all the possible C++ standards violations that could show up (more reading and tests would be necessary to understand that) I vote a strong "no" on this change.

In short:
- Hard "no" to `set(CMAKE_CXX_VISIBILITY_PRESET hidden)` (tedious to update code)
- Hard "no" to `set(CMAKE_VISIBILITY_INLINES_HIDDEN ON)` (violates C++ standard)
- Yes to testing `-fno-semantic-interposition` with GCC since Clang already does that anyway, so it can't be worse than what we currently have :P However, it should be tested that profilers and debuggers work correctly both with shared and static libs.

**Comment by nilsvu** [comment]
Here's a relevant section in the pybind11 docs: https://pybind11.readthedocs.io/en/stable/faq.html#how-can-i-create-smaller-binaries

**Comment by nilsvu** [comment]
More info on the topic in Boost:

- Boost 1.69 sets `visibility=hidden` by default: https://www.boost.org/users/history/version_1_69_0.html
- The PR that enabled it: https://github.com/boostorg/boost/pull/190
- Boost uses these macros: https://www.boost.org/doc/libs/master/libs/config/doc/html/boost_config/boost_macro_reference.html#boost_config.boost_macro_reference.macros_for_libraries_with_separate_source_code.macros_controlling_shared_library_symbol_visibility
- The macros are defined here: GCC https://github.com/boostorg/config/blob/eb3a0a98ae53ec1b35d8e94c912386e736a2ae39/include/boost/config/compiler/gcc.hpp#L110, clang https://github.com/boostorg/config/blob/eb3a0a98ae53ec1b35d8e94c912386e736a2ae39/include/boost/config/compiler/clang.hpp#L114. Note that `BOOST_SYMBOL_IMPORT` is only needed for Windows support.

**Comment by knelli2** [comment]
May be worth investigating in the future, but not enough time at the moment.
